File: android/src/main/java/expo/modules/mapboxnavigation/NavigationSessionRegistry.kt
package expo.modules.mapboxnavigation

import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

internal object NavigationSessionRegistry {
  private val lock = ReentrantLock()
  private var owner: String? = null
  private val stopHandlers = mutableMapOf<String, () -> Unit>()
  private val resumeCameraHandlers = mutableMapOf<String, () -> Unit>()
  private val cameraFollowingProviders = mutableMapOf<String, () -> Boolean>()
  private val advanceLegHandlers = mutableMapOf<String, () -> Boolean>()

  fun acquire(newOwner: String): Boolean {
    return lock.withLock {
      val current = owner
      if (current != null && current != newOwner) {
        return@withLock false
      }
      owner = newOwner
      true
    }
  }

  fun release(releasingOwner: String) {
    lock.withLock {
      if (owner == releasingOwner) {
        owner = null
      }
      stopHandlers.remove(releasingOwner)
      resumeCameraHandlers.remove(releasingOwner)
      cameraFollowingProviders.remove(releasingOwner)
      advanceLegHandlers.remove(releasingOwner)
    }
  }

  fun registerStopHandler(owner: String, handler: () -> Unit) {
    lock.withLock {
      stopHandlers[owner] = handler
    }
  }

  fun registerResumeCameraFollowingHandler(owner: String, handler: () -> Unit) {
    lock.withLock {
      resumeCameraHandlers[owner] = handler
    }
  }

  fun registerCameraFollowingProvider(owner: String, provider: () -> Boolean) {
    lock.withLock {
      cameraFollowingProviders[owner] = provider
    }
  }

  fun registerAdvanceLegHandler(owner: String, handler: () -> Boolean) {
    lock.withLock {
      advanceLegHandlers[owner] = handler
    }
  }

  fun requestStopCurrent(): Boolean {
    val handler = lock.withLock {
      val current = owner ?: return false
      stopHandlers[current]
    } ?: return false
    handler.invoke()
    return true
  }

  fun requestResumeCameraFollowingCurrent(): Boolean {
    val handler = lock.withLock {
      val current = owner ?: return false
      resumeCameraHandlers[current]
    } ?: return false
    handler.invoke()
    return true
  }

  fun requestAdvanceLegCurrent(): Boolean {
    val handler = lock.withLock {
      val current = owner ?: return false
      advanceLegHandlers[current]
    } ?: return false
    return handler.invoke()
  }

  /**
   * Whether an embedded navigation session currently holds the registry.
   *
   * Backs `getNavigationSettings().isNavigating`, which previously reported a
   * hardcoded `false`.
   */
  fun isSessionActive(): Boolean = lock.withLock { owner != null }

  fun isCurrentCameraFollowing(): Boolean {
    val provider = lock.withLock {
      val current = owner ?: return true
      cameraFollowingProviders[current]
    } ?: return true
    return provider.invoke()
  }
}

File: ios/NavigationSessionRegistry.swift
import Foundation

final class NavigationSessionRegistry {
  static let shared = NavigationSessionRegistry()

  private let lock = NSLock()
  private var owner: String?
  private var stopHandlers: [String: () -> Void] = [:]
  private var resumeCameraHandlers: [String: () -> Void] = [:]
  private var cameraFollowingProviders: [String: () -> Bool] = [:]
  private var advanceLegHandlers: [String: () -> Bool] = [:]

  private init() {}

  func acquire(owner newOwner: String) -> Bool {
    lock.lock()
    defer { lock.unlock() }

    if let current = owner, current != newOwner {
      return false
    }
    owner = newOwner
    return true
  }

  func release(owner releasingOwner: String) {
    lock.lock()
    defer { lock.unlock() }

    stopHandlers.removeValue(forKey: releasingOwner)
    resumeCameraHandlers.removeValue(forKey: releasingOwner)
    cameraFollowingProviders.removeValue(forKey: releasingOwner)
    advanceLegHandlers.removeValue(forKey: releasingOwner)
    guard owner == releasingOwner else {
      return
    }
    owner = nil
  }

  func registerStopHandler(owner: String, handler: @escaping () -> Void) {
    lock.lock()
    defer { lock.unlock() }
    stopHandlers[owner] = handler
  }

  func registerResumeCameraFollowingHandler(owner: String, handler: @escaping () -> Void) {
    lock.lock()
    defer { lock.unlock() }
    resumeCameraHandlers[owner] = handler
  }

  func registerCameraFollowingProvider(owner: String, provider: @escaping () -> Bool) {
    lock.lock()
    defer { lock.unlock() }
    cameraFollowingProviders[owner] = provider
  }

  func registerAdvanceLegHandler(owner: String, handler: @escaping () -> Bool) {
    lock.lock()
    defer { lock.unlock() }
    advanceLegHandlers[owner] = handler
  }

  func requestStopCurrent() -> Bool {
    lock.lock()
    let current = owner
    let handler = current.flatMap { stopHandlers[$0] }
    lock.unlock()
    guard let handler else {
      return false
    }
    handler()
    return true
  }

  func requestResumeCameraFollowingCurrent() -> Bool {
    lock.lock()
    let current = owner
    let handler = current.flatMap { resumeCameraHandlers[$0] }
    lock.unlock()
    guard let handler else {
      return false
    }
    handler()
    return true
  }

  func requestAdvanceLegCurrent() -> Bool {
    lock.lock()
    let current = owner
    let handler = current.flatMap { advanceLegHandlers[$0] }
    lock.unlock()
    guard let handler else {
      return false
    }
    return handler()
  }

  /// Whether an embedded navigation session currently holds the registry.
  ///
  /// Backs `getNavigationSettings().isNavigating`, which previously reported a
  /// hardcoded `false`.
  func isSessionActive() -> Bool {
    lock.lock()
    defer { lock.unlock() }
    return owner != nil
  }

  func isCurrentCameraFollowing() -> Bool {
    lock.lock()
    let current = owner
    let provider = current.flatMap { cameraFollowingProviders[$0] }
    lock.unlock()
    guard let provider else {
      return true
    }
    return provider()
  }
}
